urld829b80034_wjch111_PTSP2013Wang_tgz-pJ8-controllers_momcts_momctsCore_TestPlannerJ8/src/controllers/momcts/momctsCore/MOUCT.java

```diff
@@ -0,0 +1,514 @@
+	 * @param a		The action whose RAVE is to be updated
+	 * @param r
+	 * @param discount
+	 */
+	public void incrementRAVE(int a, String rwdType, double r, double discount){
+		incrementRwd(rwdType+"_RAVE"+a, r, discount);
+	}
+	
+	public void incrementRAVE(int a, String rwdType, double r){
+		incrementRwd(rwdType+"_RAVE"+a, r, 1.0);
+	}
+	
+	public double getRAVE(int a, String rwdType){
+		return avgR(rwdType+"_RAVE"+a);
+	}
+	
+	public double getRAVEnb(int a, String rwdType){
+		return nbs.get(rwdType+"_RAVE"+a);
+	}
+	
+	/**
+	 * Adding new new node with action = act, and inherits the context of the current node if indicated
+	 * @param act	action of the created son
+	 * @param contextInherit	indicate if the context of the current node should be inherited by the son, by default, true
+	 * @notice the act added should not be included among the existing sons, if so, return null
+	 * @return the created son
+	 */
+	public MOUCT addSon(int act, int initNb, boolean contextInherit){
+		if(hasSon(act)) return null;
+		MOUCT s = new MOUCT(act);
+		if(initNb != -1) s.setNb(defRwdTypes.firstElement(), initNb);
+		
+		if(contextInherit) {
+			@SuppressWarnings("unchecked")
+			Vector<Integer> newContext = (Vector<Integer>)context.clone();
+			newContext.add(action);
+			s.setContext(newContext);
+		}
+		sons.add(s);
+		return s;
+	}
+	
+	public MOUCT addSon(int act, int initNb){
+		return addSon(act, initNb, true);
+	}
+	
+	public MOUCT addSon(int act){
+		return addSon(act,-1);
+	}
+	
+	public Vector<Integer> sonActions(){
+		Vector<Integer> acts = new Vector<Integer>();
+		for(int i=0; i<sons.size();i++){
+			acts.add(sons.get(i).getAction());
+		}
+		return acts;
+	}
+	
+	public MOUCT getSon(int index){
+		return sons.get(index);
+	}
+	
+	/**
+	 * Find out the son whose action == act
+	 * @param act
+	 * @return the found son, if not found, return null
+	 */
+	public MOUCT findSon(int act){
+		for(MOUCT s:sons){
+			if(s.getAction() == act) return s;
+		}
+		return null;
+	}
+	
+	public boolean hasSon(int act){
+		if(findSon(act) != null) return true;
+		else return false;
+	}	
+
+	public boolean isLeaf(){
+		return !(sons.size()>0);
+	}
+	
+	public boolean isRoot(){
+		return context.size()==0;
+	}
+	
+	public int size(){
+		int sz = 1;
+		for(MOUCT s : sons){
+			sz += s.size();
+		}
+		return sz;
+	}
+	
+	/**
+	 * Show the subtree starting from the current node
+	 * @param limit		the number of levels to be shown
+	 * @param rwdTypes	indicate the reward types to be shown
+	 * @param nbType	indicate the number of visits recored in one nbs key
+	 * @param depth		pass the depth of the tree (used in presentation)
+	 * @param avg		indicate whether the average reward or the total reward is to be displayed
+	 */
+	public void showSelf(int limit, Vector<String> rwdTypes, boolean avg, String nbType, int depth){
+		if(rwdTypes.size()==0) rwdTypes = defRwdTypes;		
+		if(nbType == "") nbType = MOUCT.defRwdTypes.firstElement();//rwdTypes.firstElement();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.